### ISSUE TITLE
a series of error occurred using hg19 genome

### ISSUE THREAD

**Issue by gigglypear** [spcdot/CopyscAT#2]
Hello again!

I loaded the hg19 genome and kept other input as default to run the analysis. Some warning occurred for the generateReferences function call, and some `dplyr::mutate` error occurred for the collapseChrom3N function call.

To generate references I called:
`generateReferences(BSgenome.Hsapiens.UCSC.hg19, genomeText = "hg19", tileWidth = 1e6,
                   outputDir = "~/CopyscAT_analysis/H7")`

```
Warning message:
In `[<-.factor`(`*tmp*`, centromeres, value = "cen") :
  invalid factor level, NA generated
```

Error message for calling collapseChrom3N function:

```
Error: Problem with `mutate()` input `chromArm`.
x Input `chromArm` can't be recycled to size 3152.
ℹ Input `chromArm` is `scCNVCaller$cytoband_data$V4`.
ℹ Input `chromArm` must be size 3152 or 1, not 3113.
```

Is there any way I could fix the error?

**Comment by spcdot** [comment]
Looks like there was a bug in the generateReferences function when generating reference files for hg19 - I just updated the repo, give the latest version of the code a try - this version should generate the files correctly.

**Comment by gigglypear** [comment]
a different error occurred, and this happens to both hg38 and hg19:

```
Error in match(x, table, nomatch = 0L) : 
  'match' requires vector arguments
```

**Comment by spcdot** [comment]
Looks like that fix actually made things worse - the current edit I just uploaded should have fixed it.

**Comment by gigglypear** [comment]
The latest edit seems to fix the problem. Thank you so much!

**Comment by YiweiNiu** [comment]
Hi, sorry to re-open this thread. I also encounter errors when trying to prepare hg19 genome.

Here are the error messages and session information.

```r
> generateReferences(BSgenome.Hsapiens.UCSC.hg19,genomeText = "hg19",tileWidth = 1e6,outputDir = "/work/YiweiNiu#7001/RefData/CopyscAT")
[1] "Output to /work/YiweiNiu#7001/RefData/CopyscAT/hg19_1e+06"
   chrom    length
1   chr1 249250621
2   chr2 243199373
3   chr3 198022430
4   chr4 191154276
5   chr5 180915260
6   chr6 171115067
7   chr7 159138663
8   chr8 146364022
9   chr9 141213431
10 chr10 135534747
11 chr11 135006516
12 chr12 133851895
13 chr13 115169878
14 chr14 107349540
15 chr15 102531392
16 chr16  90354753
17 chr17  81195210
18 chr18  78077248
19 chr19  59128983
20 chr20  63025520
21 chr21  48129895
22 chr22  51304566
23  chrX 155270560
24  chrY  59373566
25  chrM     16571
GRanges object with 3457 ranges and 0 metadata columns:
                   seqnames          ranges strand
                      <Rle>       <IRanges>  <Rle>
     [1]               chr1       1-1000000      *
     [2]               chr1 1000001-2000000      *
     [3]               chr1 2000001-3000000      *
     [4]               chr1 3000001-4000000      *
     [5]               chr1 4000001-5000000      *
     ...                ...             ...    ...
  [3453] chr21_gl383580_alt         1-74652      *
  [3454] chr21_gl383581_alt        1-116690      *
  [3455] chr22_gl383582_alt        1-162811      *
  [3456] chr22_gl383583_alt         1-96924      *
  [3457] chr22_kb663609_alt         1-74013      *
  -------
  seqinfo: 298 sequences from an unspecified genome
[1] "fallback to UCSC REST API (manual table retrieval)"
Error in FUN(X[[i]], ...) : 
  no method or default for coercing “list” to “GRanges”
> .libPaths()
[1] "/work/YiweiNiu#7001/software/R/4.0" "/usr/local/lib/R/site-library"      "/usr/local/lib/R/library"          
> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C              LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
 [1] compiler  stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] BSgenome.Hsapiens.UCSC.hg19_1.4.3 BSgenome_1.58.0                   Biostrings_2.58.0                
 [4] XVector_0.30.0                    pheatmap_1.0.12                   CopyscAT_0.40                    
 [7] MASS_7.3-60                       jsonlite_1.7.2                    ggplot2_3.3.2                    
[10] sp_1.4-4                          rtracklayer_1.50.0                GenomicRanges_1.42.0             
[13] GenomeInfoDb_1.26.2               IRanges_2.24.1                    S4Vectors_0.28.1                 
[16] dplyr_1.0.2                       gplots_3.1.3                      tibble_3.0.4                     
[19] tidyr_1.1.2                       edgeR_3.32.1                      limma_3.46.0                     
[22] stringr_1.4.0                     mclust_6.0.0                      changepoint_2.2.4                
[25] zoo_1.8-8                         data.table_1.13.4                 igraph_1.2.6                     
[28] FNN_1.1.3                         Rtsne_0.15                        biomaRt_2.46.0                   
[31] fastcluster_1.2.3                 NMF_0.26                          Biobase_2.50.0                   
[34] BiocGenerics_0.36.0               cluster_2.1.0                     rngtools_1.5                     
[37] registry_0.5-1                    viridis_0.5.1                     viridisLite_0.3.0                

loaded via a namespace (and not attached):
 [1] colorspace_2.0-0            ellipsis_0.3.1              rprojroot_2.0.2             rstudioapi_0.13            
 [5] farver_2.0.3                bit64_4.0.5                 AnnotationDbi_1.52.0        fansi_0.4.1                
 [9] xml2_1.3.2                  codetools_0.2-16            doParallel_1.0.16           Rsamtools_2.6.0            
[13] gridBase_0.4-7              dbplyr_2.0.0                BiocManager_1.30.10         httr_1.4.2                 
[17] assertthat_0.2.1            Matrix_1.2-18               cli_2.2.0                   prettyunits_1.1.1          
[21] tools_4.0.3                 gtable_0.3.0                glue_1.4.2                  GenomeInfoDbData_1.2.4     
[25] reshape2_1.4.4              rappdirs_0.3.1              Rcpp_1.0.5                  cellranger_1.1.0           
[29] vctrs_0.3.6                 iterators_1.0.13            lifecycle_0.2.0             gtools_3.8.2               
[33] XML_3.99-0.5                zlibbioc_1.36.0             scales_1.1.1                hms_0.5.3                  
[37] MatrixGenerics_1.2.0        SummarizedExperiment_1.20.0 RColorBrewer_1.1-2          curl_4.3                   
[41] memoise_1.1.0               gridExtra_2.3               stringi_1.5.3               RSQLite_2.2.1              
[45] foreach_1.5.1               caTools_1.18.0              BiocParallel_1.24.1         rlang_0.4.9                
[49] pkgconfig_2.0.3             bitops_1.0-6                matrixStats_0.57.0          lattice_0.20-41            
[53] purrr_0.3.4                 GenomicAlignments_1.26.0    labeling_0.4.2              bit_4.0.4                  
[57] tidyselect_1.1.0            here_1.0.1                  plyr_1.8.6                  magrittr_2.0.1             
[61] R6_2.5.0                    generics_0.1.0              DelayedArray_0.16.0         DBI_1.1.0                  
[65] pillar_1.4.7                withr_2.3.0                 RCurl_1.98-1.2              crayon_1.3.4               
[69] KernSmooth_2.23-17          utf8_1.1.4                  BiocFileCache_1.14.0        progress_1.2.2             
[73] locfit_1.5-9.4              grid_4.0.3                  blob_1.2.1                  digest_0.6.27              
[77] openssl_1.4.3               munsell_0.5.0               askpass_1.1 
```

**Comment by YiweiNiu** [comment]
Hi, the above error is related to rtracklayer. See [this issue](https://github.com/lawremi/rtracklayer/issues/95).

**Comment by spcdot** [comment]
Hi Yiwei,

Thanks for pointing this one -- sounds like a similar cause to the last time the issue came up, with UCSC and TrackLayer updates breaking things. The fallback option should pull directly from UCSC using json but it wasn't working because of a change in the UCSC output format. I have updated the code and should work now.

I also included hg19-formatted reference files in a separate directory in the repo (hg19_references) as a quick start.

Let me know if that works.

Ana
